tools: update eslint to v1.x

targos opened the pull request:
This is still a WIP.

I create the PR now because there are several issues to discuss:
* The [`no-reserved-keys`](http://eslint.org/docs/rules/no-reserved-keys) rule has been removed. It is kind of replaced by [`quote-props`](http://eslint.org/docs/rules/quote-props) but to enable this rule, we need to choose how we want deal with quotes on object properties. The `consistent` option seems to be the most conservative one if the goal is to limit changes in our code.
* The [`indent`](http://eslint.org/docs/rules/indent) rule changed quite a lot and has now options to control the indentation of specific cases:
 * `SwitchCase`: I set it to 1, meaning that the `case` statements need 1 level of indent (== 2 spaces) compared to the `switch`. There are a few places where we don't follow this rule, like [here](https://github.com/nodejs/io.js/blob/d5ab92bcc1f0ddf7ea87a8322824a688dfd43bf5/test/pummel/test-fs-watch-file-slow.js#L21-L35).
 * `VariableDeclarator`: I set it to 2. This allows us to align multiple variable declarations with `var` or `let` like:
```js
var x = 0,
    y = 1,
    z = 2;
```

I haven't pushed the lib and test changes yet because I think there are some problems in eslint with the indent rule. I'll reference this PR when I file the issues.

silverwind commented:
I don't think we need `quote-props`, it's only useful in ES3 environments.

targos commented:
So we just remove `no-reserved-keys` ? I'm fine with that.

silverwind commented:
Patch doesn't seem to apply cleanly for me:

`error: tools/eslint/node_modules/debug/node_modules/ms/History.md: already exists in working directory`

Did you install the new eslint with `NODE_ENV=production`? I forgot that when updating the last time, so it might be best to `git rm` / `git add` everything in `tools/eslint` now.

targos commented:
I did this:
```
rm -rf tools/eslint
npm install eslint
mv node_modules/eslint tools/eslint
rm -rf node_modules
```

silverwind commented:
> So we just remove no-reserved-keys ? I'm fine with that.

Yeah. You could experiment with these options thought:

```json
quote-props: [2, "as-needed"]
quote-props: [2, "consistent-as-needed"]
```

silverwind commented:
Also, please remove `escape` and `unescape` from the globals, they're not needed anymore.

targos commented:
I have one remaining issue with the indent rule and I don't know if it's on our side or eslint's:

```js
exec('python -c "print 200000*\'C\'"', {maxBuffer: 1000},
     function(err, stdout, stderr) {
       assert.ok(err);
       assert.ok(/maxBuffer/.test(err.message));
     });
```

```
  5:8  error  Expected indentation of 8 characters but found 7  indent
  6:8  error  Expected indentation of 8 characters but found 7  indent
  7:7  error  Expected indentation of 6 characters but found 5  indent
```

targos commented:
> Also, please remove escape and unescape from the globals, they're not needed anymore.

done

silverwind commented:
> I have one remaining

If that's the only case, I'd say refactor to a function expression.

silverwind commented:
> npm install eslint

You could make that `NODE_ENV=production npm install eslint` so devDependencies don't get pulled. Keeps the size of our git tree low, and I don't think we're going to run ESLint unit tests and such :)

silverwind commented:
Also, I'd suggest doing `git rm -r tools/eslint && rm -rf tools/eslint` before and `git add tools/eslint/*` after, so leftover files aren't kept in the tree.

targos commented:
> You could make that NODE_ENV=production npm install eslint so devDependencies don't get pulled. 
Isn't it the default behavior ? I don't see dev deps in `tools/eslint/node_modules`

> suggest doing git rm -r tools/eslint && rm -rf tools/eslint before and git add tools/eslint/* after
I did it like that on last update, thanks for the tip !

silverwind commented:
> Isn't it the default behavior ?

You're right, I think npm did at some point install devDependencies when that variable wasn't set, but it seems fine now.

Fishrock123 commented:
Does this replace https://github.com/nodejs/io.js/pull/2205, or does https://github.com/nodejs/io.js/pull/2205 now depend on this?

targos commented:
#2205 depends on this.

Fishrock123 commented:
I'm not sure about the `switch`/`case:` thing. Do we only have non-indented version in those two tests?

targos commented:
@Fishrock123

Yes those tests are the only ones with non indented switch cases.

(Do not lose too much time reviewing until eslint is fixed upstream)

targos commented:
Updated to v1.1.0. A bunch of issues have been fixed but there are new ones :(

silverwind commented:
Still blocked on https://github.com/eslint/eslint/issues/3173 from what I followed.

targos commented:
@silverwind correct but not only that one. I'll continue to create issues on their tracker and cross-reference as I identify them.

michaelficarra commented:
@targos If they're all just bugs in the indent rule, why not disable that rule? It's obviously unreliable at this point. Just because you could get it to pass the current code base does not mean it won't get in somebody's way during future development.

silverwind commented:
There's no urgency in upgrading, 0.24 is working well enough. If we can avoid temporary disabling a rule, I'm all for waiting it out.

silverwind commented:
1.3 is out with the `indent` fixes, we should be able to upgrade now.

Fishrock123 commented:
Sounds good!

targos commented:
> 1.3 is out with the indent fixes, we should be able to upgrade now.
If only it was that easy !

I updated eslint and fixed all the issues, but I am absolutely not convinced about the changes I had to make to array literals declarations like here: https://github.com/targos/node/commit/5d87a8b9f3c2eb11aa7695e66c159da35f8e710a#diff-fa88b6c7c9eebf8aaf79f3ae489766dcL218

There is an [issue](https://github.com/eslint/eslint/issues/3493) about that in the eslint project but the maintainers seem to agree that it is the expected behavior...
I personally don't get why the first element should be indented by 4 spaces in this case

Fishrock123 commented:
Ugh seriously? Do we just need to make our own indentation ruleset or what?

silverwind commented:
This `VariableDeclarator` option is messing things up. Maybe we should bite the bullet and drop these multi-line `var` alingments altogether, which allows us to disable it.

cjihrig commented:
I'm +1 to just updating multi-line `var`s, if that's the only remaining blocker. It's extra noise, but will get us to where we want to be anyway. I don't think we should write our own indentation rule, as that would just be something else to maintain.

Fishrock123 commented:
Can someone post a diff of what going to var-per-line would look like?

targos commented:
I have a branch with it somewhere. Will do.

silverwind commented:
If one var per line solves all our upgrade issue relating to the `indent` rule, I'd say go for it and update this PR.

targos commented:
Updated with eslint@latest and using one var per line.

There are three remaining issues:
* weird error in lib/net.js: https://github.com/targos/node/commit/9e7bb7a3dcfe0ab978e8cb4e5f458cd92f12d6bf
* weird error in persistent history test: https://github.com/targos/node/commit/5826eb08b793274c2f7274088fa47b79fdbc426b
* glob patterns in `.eslintignore` seem to be broken. I get those errors:
```
./node tools/eslint/bin/eslint.js src lib test --rulesdir tools/eslint-rules --quiet

/Users/mzasso/git/forks/node/test/gc/node_modules/weak/lib/weak.js
   2:13  error  Missing semicolon  semi
   4:54  error  Missing semicolon  semi
   7:54  error  Missing semicolon  semi
   9:12  error  Missing semicolon  semi
  12:33  error  Missing semicolon  semi
  15:34  error  Missing semicolon  semi
  18:40  error  Missing semicolon  semi
  19:3   error  Missing semicolon  semi

/Users/mzasso/git/forks/node/test/tmp.2/中文目录/file.js
  1:1  error  Use the global form of "use strict"            strict
  1:2  error  Newline required at end of file but not found  eol-last
```

silverwind commented:
> glob patterns in .eslintignore seem to be broken

https://github.com/eslint/eslint/releases/tag/v1.4.1

targos commented:
I don't know what I did wrong but something must be missing in my first commit. If I do:
* `git checkout master`
* `git checkout eslint`
* `make jslint`

I get:
```
./node tools/eslint/bin/eslint.js src lib test --rulesdir tools/eslint-rules --quiet
module.js:338
    throw err;
    ^

Error: Cannot find module 'debug'
...
```

To do the upgrade I ran:
* `git rm -r tools/eslint`
* `rm -rf tools/eslint`
* `mv node_modules/eslint tools/eslint`
* `git add tools/eslint`

targos commented:
@silverwind 1.4.1 is the version I installed

silverwind commented:
Hmm, looks fine assuming `node_modules/eslint` came from npm. How about running `npm install` again inside `tools/eslint`?

targos commented:
That indeed adds the missing files but git doesn't detect them...

silverwind commented:
Maybe `git add --all`. Oh, and I suggest `NODE_ENV=production npm install` to avoid any devDependencies.

targos commented:
> Maybe git add --all.

Doesn't help :(
I also tried to split the remove and install parts in two commits but result is the same

> Oh, and I suggest NODE_ENV=production npm install to avoid any devDependencies.

Used `npm install --production` which I think does the same

silverwind commented:
Not sure what's the issue, but if you're desperate: `git add --all -f tools/eslint`

targos commented:
I found the issue. It is with the `.gitignore` file.
https://github.com/targos/node/commit/f939a8eeeadf8305375d5828c52d6a74dd791b06 fixes it.

silverwind commented:
Oh, that thing. We had that with npm not too long ago :)

silverwind commented:
Was [`one-var`](http://eslint.org/docs/rules/one-var) necessary to workaround issues? I'm not sure how I feel about that rule, I don't use it in my personal projects.

targos commented:
@silverwind well I didn't need it to workaround the indent issues because I didn't know about that rule. But technically most of the changes I had to do in https://github.com/targos/node/commit/13dfbc16bce551f6a3b0e7d4fc6152e000e7c158 and https://github.com/targos/node/commit/6ead03cea4c3b24f7d4aa96eb51e199a1da1b141 are exactly what the one-var rule is trying to enforce.

I thought it would be clearer for people to have a message like `Split initialized 'var' declarations into multiple statements` instead of just `Expected indentation of 2 space characters but found 4`. I personally don't care much and I wouldn't mind removing the one-var commits.

silverwind commented:
`one-var` helps with readability, so :+1: from me. The 167 line diff for the fix to multi-line var statements seems acceptable to me.

I'm worried about https://github.com/targos/node/commit/65067d06b0ec3545c8cbcca3516095b02f8af76a and https://github.com/targos/node/commit/d1170bd86dcd17479ec0774a7887dc4445623d2e commits, they clearly seem like a ESLint bug. Is there an issue for these yet?

silverwind commented:
Looking for feedback from @nodejs/collaborators regarding these stylistic changes:

https://github.com/targos/node/commit/7bd92157bf887960e55527a750fbe1a05c68dfd0
https://github.com/targos/node/commit/6465e05f84b78d10c98f7b6f6b49347c92d458a3

Trott commented:
@silverwind Regarding those two commits you are seeking feedback on: On the one hand, it's a lot of code churn just to make the linter happy. On the other hand, it's good to make the code consistent. On balance, I'm in favor if this is the only way to get eslint 1.x in play.

silverwind commented:
On second thought, I think we could do without the `one-var` change, it's not strictly necessary to get us to a updated eslint. Also, 1.6 is out, which might fix a few of the mentioned bugs, if you want to give it a go, @targos.

targos commented:
I will try v1.6 today

targos commented:
Updated to v1.6.0. I only fixed the issues in lib for now

silverwind commented:
Changes LGTM so far.

silverwind commented:
LGTM, CI: https://ci.nodejs.org/job/node-test-pull-request/416/

targos commented:
There are still problematic errors, I need to create issues in eslint

targos commented:
Especially for these:
```
/usr/home/iojs/build/workspace/node-linter/nodes/linter/test/sequential/test-regress-GH-4948.js
  22:5  error  Expected indentation of 26 space characters but found 4  indent
  25:5  error  Expected indentation of 26 space characters but found 4  indent
  27:3  error  Expected indentation of 24 space characters but found 2  indent
```

https://github.com/nodejs/node/blob/master/test/sequential/test-regress-GH-4948.js#L22

silverwind commented:
Looks like a case not covered in https://github.com/eslint/eslint/issues/3173. You could workaround by doing

```js
  var r = http.request('http://localhost:' + common.PORT, function(res) {
```

targos commented:
In this particular case yes, but there is this problem in other files

silverwind commented:
@targos Issue filed. How many of these cases are there?

targos commented:
Thank you for reporting it! The issue appears in 11 files for a total of 208 lines.
There is another issue that affects two files. I'm going to report it right now.

targos commented:
In fact I'm not sure it's a bug in eslint...

Consider this simplified example:
```js
const tests = [{
  a: 1
},
{
  a: 2
},
{
  a: 3
}];
```
Here is the error output:
```
  4:1  error  Expected indentation of 2 space characters but found 0  indent
  5:3  error  Expected indentation of 4 space characters but found 2  indent
  6:1  error  Expected indentation of 2 space characters but found 0  indent
  7:1  error  Expected indentation of 2 space characters but found 0  indent
  8:3  error  Expected indentation of 4 space characters but found 2  indent
  9:1  error  Expected indentation of 2 space characters but found 0  indent
```

I would change it to:
```js
const tests = [
  {
    a: 1
  },
  {
    a: 2
  },
  {
    a: 3
  }
];
```

Affected files:
https://github.com/nodejs/node/blob/master/test/parallel/test-repl-envvars.js#L11-L34
https://github.com/nodejs/node/blob/master/test/sequential/test-repl-persistent-history.js#L75-L151

silverwind commented:
Your suggestion looks cleaner to me, go for it.

targos commented:
Then we just need to wait for the resolution of https://github.com/eslint/eslint/issues/4075 and this will be good to go.

silverwind commented:
eslint@1.7.1 is out with above fix included. Let's hope it works out now!

jbergstroem commented:
Ping! Keen to see this land (allows me to output test to tap without exploding).

targos commented:
Looks like https://github.com/eslint/eslint/issues/4075 is not completely fixed:

```
./node tools/eslint/bin/eslint.js src lib test tools/eslint-rules \
	--rulesdir tools/eslint-rules --quiet

test/internet/test-dgram-broadcast-multi-process.js
  166:11  error  Expected indentation of 6 space characters but found 10  indent
  167:11  error  Expected indentation of 6 space characters but found 10  indent
  171:11  error  Expected indentation of 6 space characters but found 10  indent
  172:9   error  Expected indentation of 4 space characters but found 8   indent

test/internet/test-dgram-multicast-multi-process.js
  153:11  error  Expected indentation of 6 space characters but found 10  indent
  154:11  error  Expected indentation of 6 space characters but found 10  indent
  157:11  error  Expected indentation of 6 space characters but found 10  indent
  158:9   error  Expected indentation of 4 space characters but found 8   indent

test/internet/test-dns-ipv4.js
  155:5  error  Expected indentation of 25 space characters but found 4  indent
  156:5  error  Expected indentation of 25 space characters but found 4  indent
  157:5  error  Expected indentation of 25 space characters but found 4  indent
  159:5  error  Expected indentation of 25 space characters but found 4  indent
  164:5  error  Expected indentation of 25 space characters but found 4  indent
  165:3  error  Expected indentation of 23 space characters but found 2  indent

test/internet/test-dns-ipv6.js
  158:5  error  Expected indentation of 25 space characters but found 4  indent
  159:5  error  Expected indentation of 25 space characters but found 4  indent
  160:5  error  Expected indentation of 25 space characters but found 4  indent
  162:5  error  Expected indentation of 25 space characters but found 4  indent
  167:5  error  Expected indentation of 25 space characters but found 4  indent
  168:3  error  Expected indentation of 23 space characters but found 2  indent

test/parallel/test-dgram-bytes-length.js
  10:3  error  Expected indentation of 14 space characters but found 2  indent
  11:3  error  Expected indentation of 14 space characters but found 2  indent
  12:1  error  Expected indentation of 12 space characters but found 0  indent

test/parallel/test-fs-open-flags.js
  39:3  error  Expected indentation of 3 space characters but found 2  indent
  40:1  error  Expected indentation of 1 space character but found 0   indent

test/parallel/test-fs-readfile-fd.js
  6:5  error  Expected indentation of 2 space characters but found 4  indent
  7:5  error  Expected indentation of 2 space characters but found 4  indent

test/parallel/test-http-timeout.js
  29:7  error  Expected indentation of 29 space characters but found 6  indent
  31:7  error  Expected indentation of 29 space characters but found 6  indent
  39:7  error  Expected indentation of 29 space characters but found 6  indent
  40:5  error  Expected indentation of 27 space characters but found 4  indent

test/parallel/test-promises-unhandled-rejections.js
  241:3  error  Expected indentation of 12 space characters but found 2  indent
  242:3  error  Expected indentation of 12 space characters but found 2  indent
  243:3  error  Expected indentation of 12 space characters but found 2  indent
  244:1  error  Expected indentation of 10 space characters but found 0  indent
  249:3  error  Expected indentation of 12 space characters but found 2  indent
  250:3  error  Expected indentation of 12 space characters but found 2  indent
  251:3  error  Expected indentation of 12 space characters but found 2  indent
  254:1  error  Expected indentation of 10 space characters but found 0  indent
  367:3  error  Expected indentation of 12 space characters but found 2  indent
  368:3  error  Expected indentation of 12 space characters but found 2  indent
  369:3  error  Expected indentation of 12 space characters but found 2  indent
  373:3  error  Expected indentation of 12 space characters but found 2  indent
  376:1  error  Expected indentation of 10 space characters but found 0  indent
  389:3  error  Expected indentation of 12 space characters but found 2  indent
  390:3  error  Expected indentation of 12 space characters but found 2  indent
  391:3  error  Expected indentation of 12 space characters but found 2  indent
  396:3  error  Expected indentation of 12 space characters but found 2  indent
  397:3  error  Expected indentation of 12 space characters but found 2  indent
  400:1  error  Expected indentation of 10 space characters but found 0  indent
  521:3  error  Expected indentation of 12 space characters but found 2  indent
  522:3  error  Expected indentation of 12 space characters but found 2  indent
  524:3  error  Expected indentation of 12 space characters but found 2  indent
  536:1  error  Expected indentation of 10 space characters but found 0  indent
  619:3  error  Expected indentation of 12 space characters but found 2  indent
  620:3  error  Expected indentation of 12 space characters but found 2  indent
  621:3  error  Expected indentation of 12 space characters but found 2  indent
  624:3  error  Expected indentation of 12 space characters but found 2  indent
  637:1  error  Expected indentation of 10 space characters but found 0  indent

test/pummel/test-exec.js
  22:3  error  Expected indentation of 7 space characters but found 2  indent
  32:1  error  Expected indentation of 5 space characters but found 0  indent

✖ 61 problems (61 errors, 0 warnings)

Makefile:509: recipe for target 'jslint' failed
```

targos commented:
I think the common thing here is that the `function` is the first thing on the line.

targos commented:
I fixed the issues on our side.

targos commented:
CI: https://ci.nodejs.org/job/node-test-commit/889/

trevnorris commented:
I don't agree that is a style issue. For example in `test-http-timeout.js` the next aligned line is the second argument. That is exactly as it should be. Just because it starts with function doesn't mean it should be aligned differently.

targos commented:
@trevnorris I agree with you. IMO it's a bug in eslint. But they have a lot of trouble with the indent rule. Almost each time they fix something, it creates new problems...

targos commented:
Yo can see that I didn't follow the bad recommendation. Instead I forced the function to not be first in the line: https://github.com/targos/node/commit/57407eac6ab9803f5a16e2fbd918b846c148eca3#diff-bf65aca8d12f9539509f02308b0d8013R26

silverwind commented:
Here we go again: https://github.com/eslint/eslint/issues/4206

silverwind commented:
So the gist of that issue is that we should adapt our style. Maybe the cases with object literals can be taken care of too, but I'd be fine if we postpone these changes to after 1.8.x when eslint will likely break them again (https://github.com/eslint/eslint/issues/4174).

targos commented:
* rebased on master
* updated eslint to 1.7.2
* changed functions in multiline arguments after the discussion in https://github.com/eslint/eslint/issues/4206

targos commented:
now updated to v1.7.3 which has other fixes to the indent rule. I did not have to make new changes with this version.
PTAL

targos commented:
CI: https://ci.nodejs.org/job/node-test-commit/915/

silverwind commented:
LGTM, I think this is good to land now!

Fishrock123 commented:
I'm just going to cc @nodejs/tsc here because of the sheer amount of style correction churn this makes, although most of that is actually in the tests.

trevnorris commented:
If we're creating a lot of churn already then might as well switch over to `const` in as many places as makes sense.

silverwind commented:
@trevnorris I think that can be done in a seperate commit/pr, right? A few things now depend on an updated ESLint, and I wouldn't want to delay the landing of this further.

trevnorris commented:
@silverwind and add more churn? that'll just be one more step when I'm running `git blame` to see who actually made a change.

silverwind commented:
Well, the relevant rules here would be `prefer-const` and `no-const-assign`, thought I'm not really getting it if it makes any difference if a churn commit happens now or later. Or would you prefer the churn in as few commits as possible, @trevnorris?

misterdjules commented:
@trevnorris @Fishrock123 When trying to determine who made a change and when `git blame` shows me a commit that is not what I'm looking for (e.g formatting change), I use `git log -S` (with or without `--pickage-regex`) or `git log -G`.

Would that make finding what commit introduced a given change easier even after commits such as these eslint updates/fixes land?

targos commented:
@trevnorris I can do the changes but just to be clear, are you only talking about `const` (not `let`) and only for the lines that I already changed (so there will be a mix of `var` and `const` requires in some files) ?

targos commented:
* rebased
* updated eslint to 1.8.0
* used const everywhere it made sense

targos commented:
Updated eslint to 1.9.0.
cc @nodejs/collaborators

Fishrock123 commented:
The more lines this changes the less likely it will be that it ever lands...

bnoordhuis commented:
Maybe make eslint a little less overzealous?  This is a lot of basically inconsequential churn.  Code archaeologists will curse these commits many years from now.

silverwind commented:
> The more lines this changes the less likely it will be that it ever lands...

It's kind of funny, because @trevnorris suggested concentrating more churn in a single PR. It's pretty much a one-time change with a big diff, future updates will be much smoother.

> Maybe make eslint a little less overzealous?

The `indent` rule is unfortunately pretty strict, and we had a discussion with the authors where it was decided that we best follow (at least to some extent) the Google styling rules. We could disable the rule, but that would lead to a massive alignment mistakes in the long run. 

> Code archaeologists will curse these commits many years from now.

Isn't this more of a tooling issue in itself? I guess a tool could be made that 'hides' these commits.

I'll have a look at this soon.

silverwind commented:
LGTM, a 174 line diff in `lib` looks acceptable to me. Wouldn't you agree @Fishrock123?

We should run this through CI once its back up as there could be behavorial changes with `var` -> `const`.

Fishrock123 commented:
_shrug_

The test diff is... large.

silverwind commented:
It is, but churn in `test` is almost irrelevant, I'd say.

misterdjules commented:
@silverwind @bnoordhuis 

> > Code archaeologists will curse these commits many years from now.

> Isn't this more of a tooling issue in itself? I guess a tool could be made that 'hides' these commits.

`git log -S` (with or without `--pickaxe-regex`), and `git log -G` are usually enough to find the commit(s) that changed a given portion of the code while ignoring "churn" commits. That's what I use and I have yet to find a use case where it would not work well enough to prevent me from finding the commit responsible for a change.

trevnorris commented:
@misterdjules You say that as if other commands such as `git blame` aren't useful. Which this type of churn commit makes more annoying to use.

Trott commented:
If it's just the test churn that's of objectionably Brobdingnangian proportions, we can turn off the particularly churn-heavy tests in `test/.eslintrc` and turn them on again over time as tests get refactored and stuff just sort of happens by attrition (`const` replaces `var` when appropriate, etc.).

misterdjules commented:
@trevnorris This is definitely not what I'm saying.

What I'm saying is: it might be counter-productive to reject changes like this if:

1. Our only objection is our inability to use `git blame` to identify the commit(s) associated to any change in the code base
2. There are other tools as widely available and as easy to use that give us the same ability to identify the commit(s) responsible for a change.

Fishrock123 commented:
@targos something is pretty badly up with the install of eslint's deps. Maybe you could try removing the node_modules and reinstalling with npm 3?

Check `git diff 025efe7^ 025efe7 --dirstat=0.01`. `es5-ext` is installed multiple times and has way too many tests.

Relatedly, we should try to un-check-in the tests, if possible. Maybe install with `--production`?

targos commented:
I did install eslint with npm3 and the `--production` flag.
The problem here is that a lot of modules don't have the `test` directory in `.npmignore`...
I will try to add a rule in our `.gitignore` to correct this.

silverwind commented:
Maybe something like [this](https://github.com/inikulin/dmn) can help.

Fishrock123 commented:
Argh really? Multiple deps have a same dep of different versions that is figging huge? *siiiigh*

Maybe shrinkwrap would help?

targos commented:
@silverwind good find ! dmn removed 25% of the files.

So here is the procedure I followed to do the update:
```sh
# Update npm and dmn
npm install -g npm@latest
npm install -g dmn@latest
# Search for latest eslint version
npm info eslint | grep tarball
# Download and extract eslint
wget http://registry.npmjs.org/eslint/-/eslint-1.9.0.tgz
tar xf eslint-1.9.0.tgz
# Install dependencies and remove useless stuff
cd package
npm install --production
dmn clean -f
# Replace eslint with fresh installation
cd ..
rm -rf tools/eslint
mv package tools/eslint
git add tools/eslint
git commit -m "tools: update eslint to v1.9.0"
```

@Fishrock123 I ran your git diff command again (with the new SHA) and it still outputs the same, but I'm sure the test directories are gone and `es5-ext` is installed only once:
```
> find tools/eslint -name es5-ext
tools/eslint/node_modules/es5-ext
> find tools/eslint -name test
# nothing
```

targos commented:
And here is the "physical" dependency tree:
```
eslint
eslint/node_modules/amdefine
eslint/node_modules/ansi-escapes
eslint/node_modules/ansi-regex
eslint/node_modules/ansi-styles
eslint/node_modules/argparse
eslint/node_modules/array-union
eslint/node_modules/array-uniq
eslint/node_modules/arrify
eslint/node_modules/async
eslint/node_modules/balanced-match
eslint/node_modules/brace-expansion
eslint/node_modules/camelcase
eslint/node_modules/chalk
eslint/node_modules/cli-cursor
eslint/node_modules/cli-width
eslint/node_modules/code-point-at
eslint/node_modules/concat-map
eslint/node_modules/concat-stream
eslint/node_modules/core-util-is
eslint/node_modules/d
eslint/node_modules/debug
eslint/node_modules/decamelize
eslint/node_modules/deep-is
eslint/node_modules/del
eslint/node_modules/doctrine
eslint/node_modules/doctrine/node_modules/esutils
eslint/node_modules/es5-ext
eslint/node_modules/es6-iterator
eslint/node_modules/es6-map
eslint/node_modules/es6-set
eslint/node_modules/es6-symbol
eslint/node_modules/es6-weak-map
eslint/node_modules/es6-weak-map/node_modules/es6-iterator
eslint/node_modules/es6-weak-map/node_modules/es6-symbol
eslint/node_modules/escape-string-regexp
eslint/node_modules/escope
eslint/node_modules/escope/node_modules/estraverse
eslint/node_modules/espree
eslint/node_modules/esrecurse
eslint/node_modules/esrecurse/node_modules/estraverse
eslint/node_modules/estraverse
eslint/node_modules/estraverse-fb
eslint/node_modules/esutils
eslint/node_modules/event-emitter
eslint/node_modules/exit-hook
eslint/node_modules/fast-levenshtein
eslint/node_modules/figures
eslint/node_modules/file-entry-cache
eslint/node_modules/flat-cache
eslint/node_modules/generate-function
eslint/node_modules/generate-object-property
eslint/node_modules/glob
eslint/node_modules/globals
eslint/node_modules/globby
eslint/node_modules/graceful-fs
eslint/node_modules/handlebars
eslint/node_modules/has-ansi
eslint/node_modules/inflight
eslint/node_modules/inherits
eslint/node_modules/inquirer
eslint/node_modules/is-fullwidth-code-point
eslint/node_modules/is-my-json-valid
eslint/node_modules/is-path-cwd
eslint/node_modules/is-path-in-cwd
eslint/node_modules/is-path-inside
eslint/node_modules/is-property
eslint/node_modules/is-resolvable
eslint/node_modules/isarray
eslint/node_modules/js-yaml
eslint/node_modules/js-yaml/node_modules/esprima
eslint/node_modules/json-stable-stringify
eslint/node_modules/jsonify
eslint/node_modules/jsonpointer
eslint/node_modules/levn
eslint/node_modules/lodash
eslint/node_modules/lodash._arraycopy
eslint/node_modules/lodash._arrayeach
eslint/node_modules/lodash._arraymap
eslint/node_modules/lodash._baseassign
eslint/node_modules/lodash._baseclone
eslint/node_modules/lodash._basecopy
eslint/node_modules/lodash._basedifference
eslint/node_modules/lodash._baseflatten
eslint/node_modules/lodash._basefor
eslint/node_modules/lodash._baseindexof
eslint/node_modules/lodash._bindcallback
eslint/node_modules/lodash._cacheindexof
eslint/node_modules/lodash._createassigner
eslint/node_modules/lodash._createcache
eslint/node_modules/lodash._getnative
eslint/node_modules/lodash._isiterateecall
eslint/node_modules/lodash._pickbyarray
eslint/node_modules/lodash._pickbycallback
eslint/node_modules/lodash.clonedeep
eslint/node_modules/lodash.isarguments
eslint/node_modules/lodash.isarray
eslint/node_modules/lodash.isplainobject
eslint/node_modules/lodash.istypedarray
eslint/node_modules/lodash.keys
eslint/node_modules/lodash.keysin
eslint/node_modules/lodash.merge
eslint/node_modules/lodash.omit
eslint/node_modules/lodash.restparam
eslint/node_modules/lodash.toplainobject
eslint/node_modules/minimatch
eslint/node_modules/minimist
eslint/node_modules/mkdirp
eslint/node_modules/ms
eslint/node_modules/mute-stream
eslint/node_modules/number-is-nan
eslint/node_modules/object-assign
eslint/node_modules/once
eslint/node_modules/onetime
eslint/node_modules/optimist
eslint/node_modules/optionator
eslint/node_modules/os-homedir
eslint/node_modules/path-is-absolute
eslint/node_modules/path-is-inside
eslint/node_modules/pify
eslint/node_modules/pinkie
eslint/node_modules/pinkie-promise
eslint/node_modules/prelude-ls
eslint/node_modules/process-nextick-args
eslint/node_modules/read-json-sync
eslint/node_modules/read-json-sync/node_modules/graceful-fs
eslint/node_modules/readable-stream
eslint/node_modules/readline2
eslint/node_modules/restore-cursor
eslint/node_modules/rimraf
eslint/node_modules/run-async
eslint/node_modules/rx-lite
eslint/node_modules/shelljs
eslint/node_modules/source-map
eslint/node_modules/sprintf-js
eslint/node_modules/string_decoder
eslint/node_modules/strip-ansi
eslint/node_modules/strip-json-comments
eslint/node_modules/supports-color
eslint/node_modules/text-table
eslint/node_modules/through
eslint/node_modules/to-double-quotes
eslint/node_modules/to-single-quotes
eslint/node_modules/tryit
eslint/node_modules/type-check
eslint/node_modules/typedarray
eslint/node_modules/uglify-js
eslint/node_modules/uglify-js/node_modules/async
eslint/node_modules/uglify-js/node_modules/source-map
eslint/node_modules/uglify-to-browserify
eslint/node_modules/user-home
eslint/node_modules/util-deprecate
eslint/node_modules/window-size
eslint/node_modules/wordwrap
eslint/node_modules/wrappy
eslint/node_modules/write
eslint/node_modules/xml-escape
eslint/node_modules/xtend
eslint/node_modules/yargs
eslint/node_modules/yargs/node_modules/wordwrap
```

Fishrock123 commented:
@targos looks better! What's it say if you run a `git diff --shortstat` on it? (I think that is the correct command..)

targos commented:
`5061 files changed, 163286 insertions(+), 180366 deletions(-)`

Fishrock123 commented:
Wow, still a ton. :/

silverwind commented:
Thanks @targos for the v1.10 update, still LGTM.

CI: https://ci.nodejs.org/job/node-test-pull-request/849/

Is there anyone who stricly objects this 175 line 'churn'?

trevnorris commented on lib/net.js:
```diff
@@ -1130,8 +1130,12 @@ function _listen(handle, backlog) {
+  port,
+  addressType,
+  fd
+) {
```
is this a specific case where the final `)` is supposed to be dropped to the next line? b/c this isn't happening most everywhere else.

silverwind replied:
No specific reason I think, except the 4-space indent is invalid there (should be 2).

silverwind commented:
Another reason to upgrade: https://github.com/nodejs/node/pull/4612#issuecomment-170450736

Given that there don't seem to be strong opinions regarding the churn, I'd say we go for it now. @targos care to rebase once more? I don't see a particular reason to wait on `eslint@2.x` by the way, so 1.10.x should be fine.

targos commented:
>  @targos care to rebase once more

No problem. I'll try to do it this week.

targos commented:
done

silverwind commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/1221/

silverwind commented:
LGTM pending CI.

jbergstroem commented:
The linter went from 45 seconds to 1min 18sec for a run.

silverwind commented:
> The linter went from 45 seconds to 1min 18sec for a run.

Not too bad, I guess. eslint is under heavy development afte all and they have to take care of a lot of edge cases and ES6 additions.

silverwind commented:
@targos two Windows failures, not sure if flaky. The first test was modified here, the second was not:

````
not ok 73 test-fs-realpath.js
# TIMEOUT
# start tests
# fs.symlinkSync("c:\\workspace\\node-test-binary-windows\\RUN_SUBSET\\0\\VS_VERSION\\vs2013\\label\\win2008r2\\test\\fixtures/nested-index/one", "c:\\workspace\\node-test-binary-windows\\RUN_SUBSET\\0\\VS_VERSION\\vs2013\\label\\win2008r2\\test\\tmp.0/symlink", "dir")
# test_simple_relative_symlink
# fs.symlinkSync("../tmp.0/cycles/root.js", "c:\\workspace\\node-test-binary-windows\\RUN_SUBSET\\0\\VS_VERSION\\vs2013\\label\\win2008r2\\test\\tmp.0/symlink", "file")
# test_simple_absolute_symlink
# using type=dir
# test_deep_relative_file_symlink
# test_deep_relative_dir_symlink
# test_cyclic_link_protection

not ok 171 test-require-long-path.js
# internal/module.js:42
# if (content.charCodeAt(0) === 0xFEFF) {
# ^
# 
# TypeError: Cannot read property 'charCodeAt' of undefined
# at Object.stripBOM (internal/module.js:42:14)
# at Object.Module._extensions..js (module.js:432:34)
# at Module.load (module.js:367:32)
# at Function.Module._load (module.js:324:12)
# at Module.require (module.js:377:17)
# at require (internal/module.js:16:19)
# at Object.<anonymous> (c:\workspace\node-test-binary-windows\RUN_SUBSET\3\VS_VERSION\vs2015\label\win10\test\parallel\test-require-long-path.js:25:1)
# at Module._compile (module.js:423:34)
# at Object.Module._extensions..js (module.js:432:10)
# at Module.load (module.js:367:32)
````

silverwind commented:
Ah, false alarm it looks like. These failures both are present in the previous CI run:

https://ci.nodejs.org/job/node-test-binary-windows/577/

Going to land this!

jbergstroem commented:
@silverwind see https://github.com/nodejs/node/pull/4679

silverwind commented:
Landed in https://github.com/nodejs/node/commit/ed55169834a3ce16a271def9630c858626ded34d..https://github.com/nodejs/node/commit/d1aabd626428cec65e5f54c04d9e3446d1e4d3b7. Huge thanks to @targos for persisting!

jbergstroem commented:
:tada:

ChALkeR commented:
Yay!

thefourtheye commented:
@targos Great work man :-)

jasnell commented:
@rvagg @TheAlphaNerd @targos @nodejs/lts ... I'm struggling on this one to decide if it should go into LTS. If we don't land it, it's likely to make backporting changes more difficult because of the altered linting, but I'm not sure how much of a difference it will make. Any thoughts?

thefourtheye commented:
I am inclined to think that this is not an LTS material. Quoting https://github.com/nodejs/LTS#lts-plan,

> Once a release enters LTS, no new features may be added to that release. Changes are limited to bug fixes, security updates, possible npm updates, documentation updates, and certain performance improvements that can be demonstrated to not break existing applications, etc. Semver-minor changes are only permitted if required for bug fixes. Semver-major changes are only permitted if required for critical security and bug fixes.

This doesn't fall under any of the mentioned categories, right?

jasnell commented:
@thefourtheye I agree, but as we continue to evolve style rules, it'll make backporting commits a bit more difficult. I can live with it, but @TheAlphaNerd and I could definitely use a bit more active help in getting things pulled back into staging :-)

thefourtheye commented:
> as we continue to evolve style rules, it'll make backporting commits a bit more difficult.

Ah, right. I didn't think of that. Then +1 to backport this to LTS.

TheAlphaNerd commented:
+1 to backporting as well

silverwind commented:
Yeah, I think it's necessary too, or else `make test` will fail when changes which adhere to the new rules land on LTS.

rvagg commented:
I'm +1 on backporting tooling where possible, it makes it easier to manage across branches and it's not really related to API or even the code we're shipping.

targos commented:
Proposed backport: https://github.com/nodejs/node/pull/4721
